File: src/main/resources/static/javascript/ascending/QuickSortAscending.js
const fs = require('fs');
const readline = require('readline');

// QuickSort function
function quickSort(arr, low, high) {
    if (low < high) {
        const pivotIndex = partition(arr, low, high);
        quickSort(arr, low, pivotIndex - 1);
        quickSort(arr, pivotIndex + 1, high);
    }
}

// Partition function for QuickSort
function partition(arr, low, high) {
    const pivot = arr[high];
    let i = low - 1;

    for (let j = low; j < high; j++) {
        if (arr[j] <= pivot) { // Ascending order
            i++;
            [arr[i], arr[j]] = [arr[j], arr[i]];
        }
    }
    [arr[i + 1], arr[high]] = [arr[high], arr[i + 1]];

    return i + 1;
}

// Function to write an array to a CSV file
function writeToFile(filename, arr) {
    const data = arr.join(',') + '\n';
    fs.writeFileSync(filename, data);
}

// Function to get user input
function getUserInput(question) {
    return new Promise((resolve) => {
        const rl = readline.createInterface({
            input: process.stdin,
            output: process.stdout
        });
        rl.question(question, (answer) => {
            rl.close();
            resolve(answer);
        });
    });
}

async function main() {
    const n = parseInt(await getUserInput("Enter the size of the array: "), 10);
    const arr = Array.from({ length: n }, () => Math.floor(Math.random() * 100) + 1);

    const startTime = Date.now();

    const originalFile = 'original_array.csv';
    const sortedFile = 'sorted_result.csv';

    // Write the original array to file
    writeToFile(originalFile, arr);

    // Perform QuickSort
    quickSort(arr, 0, arr.length - 1);

    // Write the sorted array to file
    writeToFile(sortedFile, arr);

    const endTime = Date.now();
    const totalTime = (endTime - startTime) / 1000;

    console.log("Original and sorted arrays have been saved to CSV files.");
    console.log(`Original array file: ${originalFile}`);
    console.log(`Sorted array file: ${sortedFile}`);
    console.log(`Time taken: ${totalTime} seconds.`);
}

main();

File: src/main/resources/static/python/ascending/QuickSortAscending.py
import csv
import random
import time

def quick_sort(arr, low, high):
    if low < high:
        pivot_index = partition(arr, low, high)
        quick_sort(arr, low, pivot_index - 1)
        quick_sort(arr, pivot_index + 1, high)

def partition(arr, low, high):
    pivot = arr[high]
    i = low - 1
    for j in range(low, high):
        if arr[j] <= pivot:
            i += 1
            arr[i], arr[j] = arr[j], arr[i]
    arr[i + 1], arr[high] = arr[high], arr[i + 1]
    return i + 1

def write_to_file(filename, arr):
    with open(filename, 'w', newline='') as file:
        writer = csv.writer(file)
        writer.writerow(arr)

def main():
    n = int(input("Enter the size of the array: "))
    
    arr = [random.randint(1, 100) for _ in range(n)]

    start_time = time.time()

    # Define file names
    original_file = 'original_array.csv'
    sorted_file = 'sorted_result.csv'

    # Write the original array to file
    write_to_file(original_file, arr)

    # Perform QuickSort
    quick_sort(arr, 0, len(arr) - 1)

    # Write the sorted array to file
    write_to_file(sorted_file, arr)

    end_time = time.time()
    total_time = end_time - start_time

    print(f"Original and sorted arrays have been saved to CSV files.")
    print(f"Original array file: {original_file}")
    print(f"Sorted array file: {sorted_file}")
    print(f"Time taken: {total_time:.2f} seconds.")

if __name__ == "__main__":
    main()
